File: Greedy Algorithms/jobs.cpp
#include<iostream> 
#include<algorithm> 
using namespace std; 

// A structure to represent a job 
struct Job 
{ 
long int id;     
long int deadline; 
long int profit; 
}; 

bool comparison(Job a, Job b) 
{ 
     return (a.profit > b.profit); 
} 


void printJobScheduling(Job arr[], long int n) 
{ 
    sort(arr, arr+n, comparison); 

  long int result[n]; 
    bool slot[n];   
    for (long int i=0; i<n; i++) 
        slot[i] = false;  
    for (long int i=0; i<n; i++) 
    { 
    for (long int j=min(n, arr[i].deadline)-1; j>=0; j--) 
    { 
        if (slot[j]==false) 
        { 
            result[j] = i; 
            slot[j] = true; 
            break; 
        } 
    } 
    } 
 
    for (long int i=0; i<n; i++) 
    if (slot[i]) 
        cout << arr[result[i]].id << " "; 

  //cout << arr[result[i]].profit <<" ";

} 

int main() 
{ 
    long int x=0;
    cout<<"Enter the jobs you want  to enter";
    cin>>x;
 Job P[x];
int start=clock();
    srand(time(0)); 
cout<<start;
    for( long int i=0; i<x; i++){                         
         P[i].id=i;
        P[i].deadline=1+rand()%10;
                P[i].profit=1+rand()%100;
        cout<< "jobid are"<<P[i].id<< " " << "deadline are"<<P[i].deadline<< " " << "profit are"<<P[i].profit<<endl;
    }
    
    int y = sizeof(P) / sizeof(P[0]); 
    cout << "Following is maximum profit sequence of jobsn"; 
    printJobScheduling(P, y); 
        int end=clock();
cout<<"\n"<<(end-start)/double(CLOCKS_PER_SEC)*1000<<endl;
    return 0; 
} 

File: Greedy Algorithms/jobs.py
 
import random
def printJobScheduling(arr, t): 
	n = len(arr) 
	for i in range(n): 
		for j in range(n - 1 - i): 
			if arr[j][2] < arr[j + 1][2]: 
				arr[j], arr[j + 1] = arr[j + 1], arr[j] 
	result = [False] * t 
	job = ['-1'] * t 

	for i in range(len(arr)): 

		for j in range(min(t - 1, arr[i][1] - 1), -1, -1): 
			if result[j] is False: 
				result[j] = True
				job[j] = arr[i][0] 
				break

	print(job) 

no_of_jobs = int(input("Enter number of jobs : "))
global_jobs=[]

for i in range(0,no_of_jobs):
	local_job=[]
	local_job.append(i)
	local_job.append(1+random.randint(0, 10) ) 
	local_job.append(random.randint(10, 101) )
	global_jobs.append(local_job)

print(global_jobs)
print("Following is the sequence of jobs") 
printJobScheduling(global_jobs, no_of_jobs) 

sum=0
for i in global_jobs:
	if i[2]!=-1:
		sum= sum +i[2]

print(sum)


